`buildDotnetModule`: add support for installing pre-release tools

gaelj opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

Added an optional `allowPreRelease` argument to `buildDotnetModule` that will set the `--prerelease` flag to the `dotnet tool install` command.

This is because some useful Nuget packages have a tendency to stay in alpha release, so it's nice to be able to package them regardless.

We could also add the `--prerelease` tag unconditionally if that would be acceptable.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

corngood commented:
> We could also add the --prerelease tag unconditionally if that would be acceptable.

I can't think of a good reason not to do this in the install.  The version is pinned in the derivation.

Would specifying the version (`--version ${version}`) explicitly avoid the need for `--prerelease`?

Explicitly allowing prerelease versions could still be relevant for the update script:

```
# always skip prerelease versions for now
version=$(curl -fsSL "https://api.nuget.org/v3-flatcontainer/$nugetName/index.json" |
    jq -er '.versions | last(.[] | select(match("^[0-9]+\\.[0-9]+\\.[0-9]+$")))')
```

corngood reviewed (changes requested):
The fixup commit should be squashed.

gaelj commented:
> Would specifying the version (--version ${version}) explicitly avoid the need for --prerelease?

No, specifying an alpha version explicitly isn't enough. The pre release flag needs to be there to install a non official version or you get an error.

But I agree we should pass the pinned version to the DotNet tool installation command.

I'll also update the regex in the install script.

gaelj commented:
You are absolutely correct, @corngood : replacing `--prerelease` with  `--version [...]` does the trick.

I modified the regex in `install.sh`. I tested it by running the `curl` command in a terminal and it works as expected. Not sure how I could test the whole script though.

All commits are now squashed.

corngood commented on pkgs/build-support/dotnet/build-dotnet-global-tool/update.sh:
```diff
@@ -12,9 +12,9 @@ nixeval() {
+# allow prerelease versions
 version=$(curl -fsSL "https://api.nuget.org/v3-flatcontainer/$nugetName/index.json" |
-    jq -er '.versions | last(.[] | select(match("^[0-9]+\\.[0-9]+\\.[0-9]+$")))')
+    jq -er '.versions | last(.[] | select(match("^[0-9]+\\.[0-9]+\\.[0-9]+(-[0-9A-Za-z.-]+)?(\\+[0-9A-Za-z.-]+)?$")))')
```
I don't think we want to do this by default.  This will cause all the existing packages to update to pre-release automatically.

If we're going to do this it should be gated by something like `allowPrerelease`, but it may not be worth the trouble.  We may not want things doing automatic updates to prerelease versions at all.

corngood replied:
I think my suggestion would be to just leave this out for now.

corngood reviewed (approved):
Looks good.  Thanks!

gaelj committed:
buildDotnetModule: add support for installing pre-release tools

gaelj commented:
Force-pushed due to rebase on latest master

corngood commented:
Sorry, I should have merged this sooner.
